File: wick/step_6_cleanup.py
"""
Step 6: Cleanup / Simplification with ERI Canonicalization

Two passes:
  1. Canonicalize each integral within each tensor (sort indices, apply sign).
  2. Collect terms by canonical form and sum their numerical factors.
  3. Drop any term whose total factor is (effectively) zero.

Canonicalization rules for <p,q||r,s>:
  - Sort bra (p,q) lexicographically, applying -1 if swapped
  - Sort ket (r,s) lexicographically, applying -1 if swapped
  - Do not swap bra/ket blocks
"""

from __future__ import annotations
from itertools import permutations
import re
from typing import List
from .term import Term, Tensor, is_occ, is_virt


_EPSILON = 1e-12


def _canonicalize_integral_indices(integral_str: str) -> tuple[str, int]:
    """
    Transform '<p,q||r,s>' into canonical form with sign.
    Returns (canonical_string, sign_change).
    """
    indices = re.findall(r'[a-z0-9]+', integral_str)
    if len(indices) != 4:
        return (integral_str, 1)

    p, q, r, s = indices
    sign = 1

    # Rule 1: sort bra (p,q)
    if p > q:
        p, q = q, p
        sign *= -1

    # Rule 2: sort ket (r,s)
    if r > s:
        r, s = s, r
        sign *= -1

    return (f"<{p},{q}||{r},{s}>", sign)


def _canonicalize_tensor(tensor: Tensor) -> tuple[Tensor, int]:
    """
    Canonicalize all 4-index integrals in a tensor.
    Returns (new_tensor, total_sign).
    """
    name = tensor.name
    indices = tensor.indices
    sign = 1

    if len(indices) == 4:
        integral_str = f"<{indices[0]},{indices[1]}||{indices[2]},{indices[3]}>"
        canon_str, s = _canonicalize_integral_indices(integral_str)
        sign *= s

        canon_indices = re.findall(r'[a-z0-9]+', canon_str)
        return (Tensor(name, canon_indices), sign)

    return (tensor, sign)


def _normalize_labels(
    tensors: List[tuple[str, tuple[str, ...]]]
) -> tuple[tuple[tuple[str, tuple[str, ...]], ...], int]:
    """
    Canonicalize dummy index names so equivalent summed-index terms merge.
    """
    occ_names = iter("ijklmn")
    virt_names = iter("abcdef")
    mapping = {}

    def map_label(lbl: str) -> str:
        if lbl in mapping:
            return mapping[lbl]
        if is_occ(lbl):
            mapping[lbl] = next(occ_names)
        elif is_virt(lbl):
            mapping[lbl] = next(virt_names)
        else:
            mapping[lbl] = lbl
        return mapping[lbl]

    normalized = []
    sign = 1
    for name, idxs in tensors:
        mapped = [map_label(x) for x in idxs]
        if len(mapped) == 4:
            t, s = _canonicalize_tensor(Tensor(name, mapped))
            sign *= s
            normalized.append((name, tuple(t.indices)))
        else:
            normalized.append((name, tuple(mapped)))
    normalized.sort()
    return tuple(normalized), sign


def _term_key(term: Term) -> tuple[tuple[tuple[str, tuple[str, ...]], ...], int]:
    """
    Canonical key with sign, including permutation of occupied/virtual dummy labels.
    """
    occ_labels = sorted({idx for t in term.tensors for idx in t.indices if is_occ(idx)})
    virt_labels = sorted({idx for t in term.tensors for idx in t.indices if is_virt(idx)})

    best_key = None
    best_sign = 1

    for occ_perm in permutations(occ_labels) if occ_labels else [()]:
        occ_map = dict(zip(occ_labels, occ_perm))
        for virt_perm in permutations(virt_labels) if virt_labels else [()]:
            rename_map = dict(occ_map)
            rename_map.update(dict(zip(virt_labels, virt_perm)))

            canon_tensors = []
            total_sign = 1
            for tensor in term.tensors:
                renamed = tensor.renamed(rename_map)
                canon_tensor, s = _canonicalize_tensor(renamed)
                canon_tensors.append((canon_tensor.name, tuple(canon_tensor.indices)))
                total_sign *= s

            canon_tensors.sort()
            key, s2 = _normalize_labels(canon_tensors)
            total_sign *= s2

            if best_key is None or key < best_key or (key == best_key and total_sign > best_sign):
                best_key = key
                best_sign = total_sign

    return best_key, best_sign


def cancel_terms(terms: List[Term]) -> List[Term]:
    """Canonicalize integrals, sum identical terms, drop zeros."""
    buckets = {}   # key -> running_factor

    for term in terms:
        key, canon_sign = _term_key(term)
        effective = term.effective_factor * canon_sign

        if key in buckets:
            buckets[key] += effective
        else:
            buckets[key] = effective

    result = []
    for key, total_factor in buckets.items():
        if abs(total_factor) > _EPSILON:
            tensors = [Tensor(name, list(indices)) for name, indices in key]
            new_term = Term(abs(total_factor), 1 if total_factor > 0 else -1, tensors, [])
            result.append(new_term)

    return result

File: src/step_6_cleanup.rs
//! Step 6: Cleanup / Simplification with ERI Canonicalization
//!
//! 1. Canonicalize each integral within each tensor (sort indices, apply sign).
//! 2. Collect terms by canonical form and sum their numerical factors.
//! 3. Drop any term whose total factor is (effectively) zero.

use std::collections::HashMap;

use crate::term::{is_occ, is_virt, Tensor, Term};

const EPSILON: f64 = 1e-12;

fn canonicalize_tensor(tensor: &Tensor) -> (Tensor, i32) {
    if tensor.indices.len() != 4 {
        return (tensor.clone(), 1);
    }

    let mut p = tensor.indices[0].clone();
    let mut q = tensor.indices[1].clone();
    let mut r = tensor.indices[2].clone();
    let mut s = tensor.indices[3].clone();
    let mut sign = 1;

    if p > q {
        std::mem::swap(&mut p, &mut q);
        sign *= -1;
    }
    if r > s {
        std::mem::swap(&mut r, &mut s);
        sign *= -1;
    }

    (
        Tensor::from_strings(tensor.name.clone(), vec![p, q, r, s]),
        sign,
    )
}

type TensorKey = (String, Vec<String>);
type TermCanonKey = Vec<TensorKey>;

fn normalize_labels(tensors: &[(String, Vec<String>)]) -> (TermCanonKey, i32) {
    let mut occ_names = "ijklmn".chars();
    let mut virt_names = "abcdef".chars();
    let mut mapping: HashMap<String, String> = HashMap::new();
    let mut sign = 1;

    let map_label = |lbl: &str,
                     mapping: &mut HashMap<String, String>,
                     occ_names: &mut std::str::Chars,
                     virt_names: &mut std::str::Chars|
     -> String {
        if let Some(mapped) = mapping.get(lbl) {
            return mapped.clone();
        }
        let new_label = if is_occ(lbl) {
            occ_names.next().unwrap().to_string()
        } else if is_virt(lbl) {
            virt_names.next().unwrap().to_string()
        } else {
            lbl.to_string()
        };
        mapping.insert(lbl.to_string(), new_label.clone());
        new_label
    };

    let mut normalized: Vec<(String, Vec<String>)> = Vec::new();
    for (name, idxs) in tensors {
        let mapped: Vec<String> = idxs
            .iter()
            .map(|x| map_label(x, &mut mapping, &mut occ_names, &mut virt_names))
            .collect();
        if mapped.len() == 4 {
            let t = Tensor::from_strings(name.clone(), mapped);
            let (ct, s) = canonicalize_tensor(&t);
            sign *= s;
            normalized.push((name.clone(), ct.indices));
        } else {
            normalized.push((name.clone(), mapped));
        }
    }
    normalized.sort();
    (normalized, sign)
}

fn term_key(term: &Term) -> (TermCanonKey, i32) {
    let mut occ_labels: Vec<String> = Vec::new();
    let mut virt_labels: Vec<String> = Vec::new();

    for t in &term.tensors {
        for idx in &t.indices {
            if is_occ(idx) && !occ_labels.contains(idx) {
                occ_labels.push(idx.clone());
            }
            if is_virt(idx) && !virt_labels.contains(idx) {
                virt_labels.push(idx.clone());
            }
        }
    }
    occ_labels.sort();
    virt_labels.sort();

    let mut best_key: Option<TermCanonKey> = None;
    let mut best_sign = 1;

    let occ_perms = permutations(&occ_labels);
    let virt_perms = permutations(&virt_labels);

    for occ_perm in &occ_perms {
        let mut occ_map: HashMap<String, String> = HashMap::new();
        for (orig, perm) in occ_labels.iter().zip(occ_perm.iter()) {
            occ_map.insert(orig.clone(), perm.clone());
        }

        for virt_perm in &virt_perms {
            let mut rename_map = occ_map.clone();
            for (orig, perm) in virt_labels.iter().zip(virt_perm.iter()) {
                rename_map.insert(orig.clone(), perm.clone());
            }

            let mut canon_tensors: Vec<(String, Vec<String>)> = Vec::new();
            let mut total_sign = 1;

            for tensor in &term.tensors {
                let renamed = tensor.renamed(&rename_map);
                let (canon_tensor, s) = canonicalize_tensor(&renamed);
                canon_tensors.push((canon_tensor.name.clone(), canon_tensor.indices.clone()));
                total_sign *= s;
            }

            canon_tensors.sort();
            let (key, s2) = normalize_labels(&canon_tensors);
            total_sign *= s2;

            let replace = match &best_key {
                None => true,
                Some(bk) => key < *bk || (key == *bk && total_sign > best_sign),
            };
            if replace {
                best_key = Some(key);
                best_sign = total_sign;
            }
        }
    }

    (best_key.unwrap_or_default(), best_sign)
}

fn permutations(items: &[String]) -> Vec<Vec<String>> {
    if items.is_empty() {
        return vec![vec![]];
    }
    let mut result = Vec::new();
    for i in 0..items.len() {
        let mut rest: Vec<String> = items.to_vec();
        let item = rest.remove(i);
        for mut perm in permutations(&rest) {
            perm.insert(0, item.clone());
            result.push(perm);
        }
    }
    result
}

pub fn cancel_terms(terms: &[Term]) -> Vec<Term> {
    let mut buckets: Vec<(TermCanonKey, f64)> = Vec::new();
    let mut key_index: HashMap<TermCanonKey, usize> = HashMap::new();

    for term in terms {
        let (key, canon_sign) = term_key(term);
        let effective = term.effective_factor() * canon_sign as f64;

        if let Some(&idx) = key_index.get(&key) {
            buckets[idx].1 += effective;
        } else {
            key_index.insert(key.clone(), buckets.len());
            buckets.push((key, effective));
        }
    }

    let mut result = Vec::new();
    for (key, total_factor) in &buckets {
        if total_factor.abs() > EPSILON {
            let tensors: Vec<Tensor> = key
                .iter()
                .map(|(name, indices)| Tensor::from_strings(name.clone(), indices.clone()))
                .collect();
            let sign = if *total_factor > 0.0 { 1 } else { -1 };
            result.push(Term::new(total_factor.abs(), sign, tensors, vec![], vec![]));
        }
    }

    result
}
